clr/commands.py

```diff
@@ -120,28 +102,93 @@ def get_command_spec(cmd):
     return args, vararg
 
 class System(object):
-    ns = 'system'
+    """System namespace for the clr tool.
+
+    Commands defined here will be avaliable directly without specifying a
+    namespace. For example `clr help` instead of `clr system:help`. Be careful
+    not to define commands here that have the same name as a defined namespace
+    or it may be obscured."""
+
     descr = 'system commands'
 
-    def cmd_help(self, which=None):
+    def cmd_help(self, query=None, query2=None):
         """
-        provides help for commands, when specified, `which' can be one
+        provides help for commands, when specified, `query' can be one
         either a namespace or a namespace:command tuple.
         """
-        if which is None:
+        if not query:
             print('Available namespaces')
-            for obj in list(get_commands().values()):
-                print(' ', obj.ns.ljust(20), '-', obj.descr)
-        elif which.find(':') < 0:
-            if which in get_commands():
-                obj = get_commands()[which]
-
-                print(which, '-', obj.descr)
-
-                for k in [k for k in dir(obj) if k.startswith('cmd_')]:
-                    cmd = '%s:%s' % (which, k[4:])
-                    print_help_for_cmd(cmd, prefix='  ')
-            else:
-                print_help_for_cmd('system:%s' % which)
-        else:
-            print_help_for_cmd(which)
+            for key, namespace in get_namespaces().items():
+                print(' ', key.ljust(20), '-', namespace.descr)
+            return
+
+        # If they passed just one arg and it is a namespace key, print help for the full namespace.
+        if query in NAMESPACE_KEYS and not query2:
+            namespace = get_namespace(query)
+            descr = namespace.longdescr if hasattr(namespace, 'longdescr') else namespace.descr
+            print(query, '-', descr)
+
+            for command in list_commands(query):
+                self.print_help_for_command(query, command, prefix='  ')
+            return
+
+        if query2: query = f'{query}:{query2}'
+        _, _, namespace_key, command_name = resolve_command(query)
+        self.print_help_for_command(namespace_key, command_name)
+
+    def print_help_for_command(self, namespace_key, command_name, prefix=''):
+        w = textwrap.TextWrapper(
+            initial_indent=prefix, subsequent_indent=prefix,
+            width=70)
+
+        command = get_command(namespace_key, command_name)
+
+        spec, vararg = get_command_spec(command)
+
+        is_default = lambda a_s: a_s[1] is intern('default')
```

Comment:
I know it's copy/paste from earlier code, but I'm pretty sure that using `a_s[1] == 'default'` will work, which means we can probably ditch another import.

Reply:
So that seems like it works to me. Crazy part is the current code doesn't work. :/

Reply:
Oh I see what was happening now. Was using intern to create a sentinel value. Placed there when "creating" the spec

Reply:
I think this is better. I may do a better revamp of the reflective bits in a later PR